// rc-color-picker/lib/Panel.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x2, _x3, _x4) { var _again = true; _function: while (_again) { var object = _x2, property = _x3, receiver = _x4; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x2 = parent; _x3 = property; _x4 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _colr = require('colr');

var _colr2 = _interopRequireDefault(_colr);

var _Board = require('./Board');

var _Board2 = _interopRequireDefault(_Board);

var _Preview = require('./Preview');

var _Preview2 = _interopRequireDefault(_Preview);

var _Ribbon = require('./Ribbon');

var _Ribbon2 = _interopRequireDefault(_Ribbon);

var _Alpha = require('./Alpha');

var _Alpha2 = _interopRequireDefault(_Alpha);

var _Params = require('./Params');

var _Params2 = _interopRequireDefault(_Params);

var _utilsValidationColor = require('./utils/validationColor');

var _utilsValidationColor2 = _interopRequireDefault(_utilsValidationColor);

function noop() {}

var colr = new _colr2['default']();

var Panel = (function (_React$Component) {
  _inherits(Panel, _React$Component);

  function Panel(props) {
    var _this = this;

    _classCallCheck(this, Panel);

    _get(Object.getPrototypeOf(Panel.prototype), 'constructor', this).call(this, props);

    var color = props.color || props.defaultColor;
    var hsv = colr.fromHex(color).toHsvObject();

    this.state = {
      paramsHsv: hsv,
      hsv: hsv,
      alpha: props.alpha || props.defaultAlpha
    };

    var events = ['onChange', 'onAlphaChange', 'onFocus', 'onBlur', 'onSystemColorPickerOpen'];
    // bind methods
    events.forEach(function (m) {
      _this[m] = _this[m].bind(_this);
    });
  }

  _createClass(Panel, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.color) {
        var hsv = colr.fromHex(nextProps.color).toHsvObject();
        this.setState({
          hsv: hsv
        });
      }
      if (nextProps.alpha !== undefined) {
        this.setState({
          alpha: nextProps.alpha
        });
      }
    }
  }, {
    key: 'onChange',
    value: function onChange(hsvObject) {
      var syncParams = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

      var hsv = hsvObject;
      var state = {
        hsv: hsv
      };
      if (syncParams) {
        state.paramsHsv = hsv;
      }
      this.setState(state);

      var ret = {
        color: this.getHexColor(hsv),
        hsv: hsv,
        alpha: this.state.alpha
      };
      this.props.onChange(ret);
    }
  }, {
    key: 'onSystemColorPickerOpen',
    value: function onSystemColorPickerOpen(e) {
      // only work with broswer which support color input
      if (e.target.type === 'color') {
        this.systemColorPickerOpen = true;
      }
    }
  }, {
    key: 'onAlphaChange',
    value: function onAlphaChange(alpha) {
      if (this.props.alpha === undefined) {
        this.setState({
          alpha: alpha
        });
      }
      this.props.onChange({
        color: this.getHexColor(),
        hsv: this.state.hsv,
        alpha: alpha
      });
    }
  }, {
    key: 'onFocus',
    value: function onFocus() {
      if (this._blurTimer) {
        clearTimeout(this._blurTimer);
        this._blurTimer = null;
      } else {
        this.props.onFocus();
      }
    }
  }, {
    key: 'onBlur',
    value: function onBlur() {
      var _this2 = this;

      if (this._blurTimer) {
        clearTimeout(this._blurTimer);
      }
      this._blurTimer = setTimeout(function () {
        // if is system color picker open, then stop run blur
        if (_this2.systemColorPickerOpen) {
          _this2.systemColorPickerOpen = false;
          return;
        }

        _this2.props.onBlur();
      }, 100);
    }
  }, {
    key: 'getHexColor',
    value: function getHexColor(hsv) {
      return colr.fromHsvObject(hsv || this.state.hsv).toHex();
    }
  }, {
    key: 'render',
    value: function render() {
      var prefixCls = this.props.prefixCls;
      var hsv = this.state.hsv;
      var alpha = this.state.alpha;
      return _react2['default'].createElement(
        'div',
        {
          className: prefixCls,
          style: this.props.style,
          onFocus: this.onFocus,
          onBlur: this.onBlur,
          tabIndex: '0'
        },
        _react2['default'].createElement(
          'div',
          { className: prefixCls + '-' + 'inner' },
          _react2['default'].createElement(_Board2['default'], {
            rootPrefixCls: prefixCls,
            hsv: hsv,
            onChange: this.onChange
          }),
          _react2['default'].createElement(
            'div',
            { className: prefixCls + '-' + 'wrap' },
            _react2['default'].createElement(
              'div',
              { className: prefixCls + '-' + 'wrap-ribbon' },
              _react2['default'].createElement(_Ribbon2['default'], {
                rootPrefixCls: prefixCls,
                hsv: hsv,
                onChange: this.onChange
              })
            ),
            _react2['default'].createElement(
              'div',
              { className: prefixCls + '-' + 'wrap-alpha' },
              _react2['default'].createElement(_Alpha2['default'], {
                rootPrefixCls: prefixCls,
                alpha: alpha,
                hsv: hsv,
                onChange: this.onAlphaChange
              })
            ),
            _react2['default'].createElement(
              'div',
              { className: prefixCls + '-' + 'wrap-preview' },
              _react2['default'].createElement(_Preview2['default'], {
                rootPrefixCls: prefixCls,
                alpha: alpha,
                onChange: this.onChange,
                onInputClick: this.onSystemColorPickerOpen,
                hsv: hsv
              })
            )
          ),
          _react2['default'].createElement(
            'div',
            { className: prefixCls + '-' + 'wrap', style: { height: 40, marginTop: 6 } },
            _react2['default'].createElement(_Params2['default'], {
              rootPrefixCls: prefixCls,
              hsv: this.state.paramsHsv,
              alpha: alpha,
              onAlphaChange: this.onAlphaChange,
              onChange: this.onChange,
              mode: this.props.mode
            })
          )
        )
      );
    }
  }]);

  return Panel;
})(_react2['default'].Component);

exports['default'] = Panel;

Panel.propTypes = {
  defaultAlpha: _react.PropTypes.number,
  defaultColor: _utilsValidationColor2['default'],
  // can custom
  prefixCls: _react.PropTypes.string,
  color: _utilsValidationColor2['default'],
  alpha: _react.PropTypes.number,
  style: _react.PropTypes.object,
  onChange: _react.PropTypes.func,
  onFocus: _react.PropTypes.func,
  onBlur: _react.PropTypes.func,
  mode: _react.PropTypes.oneOf(['RGB', 'HSL', 'HSB'])
};

Panel.defaultProps = {
  prefixCls: 'rc-color-picker-panel',
  defaultColor: '#ff0000',
  defaultAlpha: 100,
  style: {},
  onChange: noop,
  onFocus: noop,
  onBlur: noop,
  mode: 'RGB'
};
module.exports = exports['default'];
